### PR TITLE
Adds 'DeviceBuilder' factory and helper 'multiDeviceGolden' method to create single golden files for multiple Device configurations and scenarios

### PR DESCRIPTION
**Pull request by moonytoes29**
**Problem statement:**

Capturing golden files for a screen that has multiple differences across device sizes and multiple complicated ui states. 

Current package says to use 'multiScreenGolden' for the same widget rendered in multiple device sizes (i.e Device.phone, Device.tabletLandscape). This however, produces a single png golden file per device. When adding multiple scenarios/states for that screen and needing still to see across multiple devices, this can grow the number of files considerably.

**Proposed solution:**

**DeviceBuilder factory**

The DeviceBuilder class allows user to create multiple scenarios across multiple devices in a single png golden file. The example added was around the default flutter counter view and capturing ui based on user taps.

```dart
  testGoldens('DeviceBuilder - multiple scenarios - with afterPump',
      (tester) async {
    final builder = DeviceBuilder()
      ..overrideDevicesForAllScenarios(devices: [
        Device.phone,
        Device.iphone11,
        Device.tabletPortrait,
        Device.tabletLandscape,
      ])
      ..addDeviceScenario(
        widget: FlutterDemoPage(),
        name: 'default page',
      )
      ..addDeviceScenario(
        widget: FlutterDemoPage(),
        name: 'tap once',
        afterPump: (scenarioWidgetKey) async {
          final finder = find.descendant(
            of: find.byKey(scenarioWidgetKey),
            matching: find.byIcon(Icons.add),
          );
          expect(finder, findsOneWidget);
          await tester.tap(finder);
        },
      )
      ..addDeviceScenario(
        widget: FlutterDemoPage(),
        name: 'tap five times',
        afterPump: (scenarioWidgetKey) async {
          final finder = find.descendant(
            of: find.byKey(scenarioWidgetKey),
            matching: find.byIcon(Icons.add),
          );
          expect(finder, findsOneWidget);

          await tester.tap(finder);
          await tester.tap(finder);
          await tester.tap(finder);
          await tester.tap(finder);
          await tester.tap(finder);
        },
      );

    await tester.pumpDeviceBuilder(builder);

    await screenMatchesGolden(tester, 'flutter_demo_page_multiple_scenarios');
  });
```

The DeviceBuilder exposes an 'afterPump' callback to perform any interactions required to render the scenario across all device sizes. For simpler widgets/screens where a ui state can be instantiated statically through construction, then the 'afterPump' functionality wouldn't be needed.

This produces a single golden file:

flutter_demo_page_multiple_scenarios

The simplest use of this doesn't require you to call `overrideDevicesForAllScenarios`, as by default it will use `GoldenToolkit.configuration.defaultDevices`. You can also omit scenario `name` parameter to. 

```dart
final builder = DeviceBuilder()..addDeviceScenario( widget: FlutterDemoPage());
```

**multiDeviceGolden**

This is a helper method much like the existing `multiScreenGolden`, except that it produces one single golden file instead of multiple per device. This method can be used if you just one a single scenario and don't need to use `DeviceBuilder` directly.

```dart
  testGoldens('multiDeviceGolden', (tester) async {
    await multiDeviceGolden(
      tester,
      'flutter_demo_multi_device_golden',
      widget: FlutterDemoPage(),
      afterPump: (scenarioWidgetKey) async {
        final finder = find.descendant(
          of: find.byKey(scenarioWidgetKey),
          matching: find.byIcon(Icons.add),
        );
        expect(finder, findsOneWidget);
        await tester.tap(finder);
      },
    );
  });
```

**Other fixes**

The current tests were not passing on latest flutter stable 1.22.1. Had to regenerate some golden files. Also had issues with analysis options when opening root folder and looking at weather_widgets.dart. So ignored some issues in there to help analysis step pass.

### CODE REVIEW

**Comment by mZadorskii** [comment]
Thanks for contributing! I have not looked into your PR yet but want to give some explanation: One of the reasons we went with single file is to speed up goldens review process. When single file has 5-10 screens and something has changed it's not trivial to review such a huge golden to make sure nothing else accidentally has changed in that file.

**Comment by coreysprague** [comment]
Thanks @moonytoes29! Thanks for taking the time to do this. At a high level, the concept seems good and I think this won't take much effort to integrate.

That being said, I have a couple high level pieces of feedback:
- the coverage bits are a little funny in this repo. The "example" project contains examples to demonstrate usage, but doesn't necessarily exhaustively test the functionality/edge cases/etc. Additionally, any tests in /example don't count for coverage in the golden_toolkit package itself. It might make sense to review your tests in example and see if some should move into the golden_toolkit package. That would probably satisfy the coverage gap.
- Since this is new functionality, I think the main readme documentation should be updated to explain this new API: why/when it should be used, etc.

I'll provide more targeted feedback on the code later today or tomorrow.

Thanks!

**Review by coreysprague — CHANGES_REQUESTED** [review]
Sorry for the delay -- it's been a busy week.

I added some feedback. 

Overall, I like the feature. It's highlighting issues with our existing APIs... I think there is a lot of opportunity for us to consolidate and simplify some of the disparate APIs that are accumulating across the package... but that's a big effort, and I don't want to make you suffer through that 😄

**Comment by coreysprague** [comment]
@moonytoes29 sorry, I had to upgrade to the latest flutter stable in order to merge another branch, so it created conflicts... hopefully easily resolvable.

File: packages/golden_toolkit/README.md [hunk 1]
```diff
@@ -109,6 +187,53 @@ await multiScreenGolden(
...
 
+### multiDeviceGolden
+
+The multiDeviceGolden assertion is used to capture a golden of a single widget that has different simulated device sizes & characteristics inside. All stitched together in one single png file. As opposed to multiple in multiScreenGolden
+
+The typical use case is to validate that a UI is responsive for both phone & tablet in one single golden file.
+
+```dart
+testGoldens('multiDeviceGolden', (tester) async {
+    await multiDeviceGolden(
```
[the remark this answers is not in the document]
Comment (moonytoes29) [reply packages/golden_toolkit/README.md:198]: Hi @coreysprague - thank you for taking the time to review this PR. Sorry i've not had a chance to reply. Been a hell of a week.

To be fair, this method was more of a change i did to help test this whole new api at work. Rather than me having to refactor all of our `multiScreenGolden` calls to use the new DeviceBuilder structure, i thought it'd be easier to do this function wrapper and a more simple find and replace.

As mentioned i the README, the same functionality definitely can be done by just using the DeviceBuilder with one scenario.

I'm happy to drop this if you'd like. Can always have it as an extension on the project side rather than package.

File: packages/golden_toolkit/lib/src/device_builder.dart [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+
+  /// Can be used to wrap all scenario widgets. Useful if you wish to
+  /// provide consistent UI treatment to all of them or need to inject dependencies.
+  final WidgetWrapper wrap;
+
+  ///  background [bgColor] color of output .png file
+  final Color bgColor;
+
+  /// list of created DeviceScenarios for each device type
+  final List<DeviceScenario> scenarios = [];
```
[the remark this answers is not in the document]
Comment (moonytoes29) [reply packages/golden_toolkit/lib/src/device_builder.dart:69]: This list is used by the `pumpDeviceBuilder` in order to call the individual `afterPump` (function name tbh 😄) for each of them. So cant make this private. Only thing i can think to do, would be to maybe expose a method on DeviceBuilder to `callAfterPumpsOnScenarios` or something like that maybe?

File: packages/golden_toolkit/lib/src/device_builder.dart [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+    this.key,
+    this.afterPump,
+    this.widget,
+  });
+
+  /// key that references created scenario and specific device
+  final Key key;
+
+  /// function that is executed after initial pump of widget
+  final AfterPump afterPump;
```
[the remark this answers is not in the document]
Comment (moonytoes29) [reply packages/golden_toolkit/lib/src/device_builder.dart:31]: I dont know if "setup" implies before. Technically this is after its rendered the frame at least once. Hmm, names are not my forte either.

What about `onCreate`?

File: packages/golden_toolkit/lib/src/device_builder.dart [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+/// https://opensource.org/licenses/BSD-3-Clause
+/// ***************************************************
+import 'package:flutter/widgets.dart';
+import 'package:meta/meta.dart';
+
+import '../golden_toolkit.dart';
+import 'testing_tools.dart';
+
+/// Class containing whats required to create a Scenario
+class DeviceScenario {
```
[the remark this answers is not in the document]
Comment (moonytoes29) [reply packages/golden_toolkit/lib/src/device_builder.dart]: updated

File: packages/golden_toolkit/lib/src/device_builder.dart [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+  /// list of created DeviceScenarios for each device type
+  final List<DeviceScenario> scenarios = [];
+
+  List<Device> _devicesForScenarios = GoldenToolkit.configuration.defaultDevices ?? [Device.phone];
+  int get _numberOfDevicesPerScenario => _devicesForScenarios?.length ?? 1;
+
+  /// Overrides the list of devices that are rendered for each scenario
+  /// Otherwise will default to using `GoldenToolkit.configuration.defaultDevices`
+  void overrideDevicesForAllScenarios({@required List<Device> devices}) {
+    if (devices != null) {
```
[the remark this answers is not in the document]
Comment (moonytoes29) [reply packages/golden_toolkit/lib/src/device_builder.dart]: updated

File: packages/golden_toolkit/lib/src/device_builder.dart [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+    }
+  }
+
+  /// [addDeviceScenario] will add a [_DeviceScenario] for each device listed
+  /// under [_devicesForScenarios]
+  ///
+  /// [widget] widget you'd like rendered ad Device sizes
+  ///
+  /// [name] name of scenario 'e.g 'error_state'
+  void addDeviceScenario({
```
[the remark this answers is not in the document]
Comment (moonytoes29) [reply packages/golden_toolkit/lib/src/device_builder.dart]: updated

File: packages/golden_toolkit/lib/src/testing_tools.dart [hunk 7]
```diff
@@ -53,6 +56,28 @@ extension TestingToolsExtension on WidgetTester {
...
+
+  /// Extension for a [WidgetTester] that pumps a [DeviceBuilder] class
+  ///
+  /// [WidgetWrapper] defaults to [materialAppWrapper]
+  ///
+  /// [textScaleSize] set's the text scale size (usually in a range from 1 to 3)
+  Future<void> pumpDeviceBuilder(
+    DeviceBuilder deviceBuilder, {
+    WidgetWrapper wrapper,
+    double textScaleSize = 1.0,
```
[the remark this answers is not in the document]
Comment (moonytoes29) [reply packages/golden_toolkit/lib/src/testing_tools.dart]: i agree, removed

File: packages/golden_toolkit/lib/src/multi_screen_golden.dart [hunk 8]
```diff
@@ -84,3 +84,44 @@ Future<void> multiScreenGolden(
...
+///
+/// [devices] list of devices to run the tests - overriding `GoldenToolkit.configuration.defaultDevices`
+///
+/// [afterPump] is passed into DeviceBuilder scenario to be called after that device specific widget is rendered
+///
+Future<void> multiDeviceGolden(
+  WidgetTester tester,
+  String name, {
+  Widget widget,
+  CustomPump customPump,
```
[the remark this answers is not in the document]
Comment (moonytoes29) [reply packages/golden_toolkit/lib/src/multi_screen_golden.dart]: removed

File: packages/golden_toolkit/README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (coreysprague) [reply packages/golden_toolkit/README.md:198]: Thanks @moonytoes29. Let's leave it as a project extension. I really do like the simplicity, but I think it might cause some confusion until we have a chance to consolidate and cleanup the APIs.

File: packages/golden_toolkit/lib/src/device_builder.dart (hunk 4 above)
[the remark this answers is not in the document]
Comment (coreysprague) [reply packages/golden_toolkit/lib/src/device_builder.dart]: 👍

File: packages/golden_toolkit/lib/src/device_builder.dart (hunk 2 above)
[the remark this answers is not in the document]
Comment (coreysprague) [reply packages/golden_toolkit/lib/src/device_builder.dart]: I am always torn on this. The OO purist in me wants to encapsulate and make a method like you suggest, but that ends up in a more opinionated API that makes it harder for consumers to tweak/work-around issues. We can go ahead and leave it I think.

File: packages/golden_toolkit/lib/src/device_builder.dart (hunk 3 above)
[the remark this answers is not in the document]
Comment (coreysprague) [reply packages/golden_toolkit/lib/src/device_builder.dart:31]: I think you are right about "setup". Let me see if anyone else on my team has a good suggestion...

File: packages/golden_toolkit/example/lib/src/weather_widgets.dart [hunk 9]
```diff
@@ -7,6 +7,9 @@
 /// ***************************************************
 // ignore_for_file: public_member_api_docs
 
+// ignore_for_file: public_member_api_docs
```
Comment (fatherOfLegends) [inline packages/golden_toolkit/example/lib/src/weather_widgets.dart:10]: This ignore is a duplicate.

File: packages/golden_toolkit/example/lib/src/weather_widgets.dart [hunk 10]
```diff
@@ -7,6 +7,9 @@
...
+// ignore_for_file: prefer_const_constructors
```
Comment (fatherOfLegends) [inline packages/golden_toolkit/example/lib/src/weather_widgets.dart:11]: What is the reason for ignoring const constructors?

File: packages/golden_toolkit/example/test/flutter_demo_page_test.dart [hunk 11]
```diff
@@ -0,0 +1,101 @@
+import 'package:flutter/material.dart';
+import 'package:flutter_test/flutter_test.dart';
+import 'package:golden_toolkit/golden_toolkit.dart';
+
+// ignore: avoid_relative_lib_imports
```
Comment (fatherOfLegends) [inline packages/golden_toolkit/example/test/flutter_demo_page_test.dart:5]: Rather than ignore this lint I thin you can just import the package.

File: packages/golden_toolkit/lib/src/device_builder.dart (hunk 3 above)
[the remark this answers is not in the document]
Comment (fatherOfLegends) [reply packages/golden_toolkit/lib/src/device_builder.dart:31]: What about something like `onEach` or `afterEach`?

File: packages/golden_toolkit/example/lib/src/weather_widgets.dart (hunk 10 above)
Comment (moonytoes29) [reply packages/golden_toolkit/example/lib/src/weather_widgets.dart:11]: i cant remember what happened. If i didnt open the workspace or something, but i believe i was getting issues with analysis saying i was missing const keywords in places. Rather than have to fix a widget/class i wasnt touching (one that should have already pased the linting), i added the prefer_const_constructors.

I've opened the workspace and removed now locally and it hasnt had any issues. So i will remove.

Interesting point i found when gathering coverage in a work project, in a test file if you construct the object your testing with a const prefix then coverage wasnt taking the fact that the constructor was covered. Removing the const keyword made that work.

Havent retested that in latest flutter stable tbh.

File: packages/golden_toolkit/example/test/flutter_demo_page_test.dart (hunk 11 above)
Comment (moonytoes29) [reply packages/golden_toolkit/example/test/flutter_demo_page_test.dart:5]: `flutter_demo_page.dart` is exported by the `example.dart` barrel file. So i will change this import to be `import 'package:example/example.dart';` and remove the ignore

**Comment by moonytoes29** [comment]
On pull requests there are some coverage level thresholds that are enforced. In another PR, it might be good to add a vs code task to match what the pipeline runs. i.e checking for line length, analysis bits and the required coverage percentage maybe?

File: packages/golden_toolkit/lib/src/device_builder.dart (hunk 3 above)
[the remark this answers is not in the document]
Comment (fatherOfLegends) [reply packages/golden_toolkit/lib/src/device_builder.dart:31]: Sorry just realized this was on an individual `_DeviceScenario`. I think `onCreate` would work as @moonytoes29 said. 
Other ideas... 
`scenario`
`userBehavior`

[the remark this answers is not in the document]
Comment (moonytoes29) [reply packages/golden_toolkit/lib/src/device_builder.dart:31]: I think afterEach isnt far off, although arguably afterInitialPump might be more accurate. 'Each' implies that it runs after each scenario, but it just runs once really. Its a tough one.

Happy to change if it helps the merge, but i think 'afterPump' is preferrable. But final say is with you guys. 😄  ?

File: packages/golden_toolkit/lib/src/multi_screen_golden.dart [hunk 12]
```diff
@@ -84,3 +84,37 @@ Future<void> multiScreenGolden(
...
+///
+/// Will output a single golden file for all devices
+///
+/// [name] is a file name output, must NOT include extension like .png
+///
+/// [devices] list of devices to run the tests - overriding `GoldenToolkit.configuration.defaultDevices`
+///
+/// [afterPump] is passed into DeviceBuilder scenario to be called after that device specific widget is rendered
+///
+Future<void> multiDeviceGolden(
```
Comment (fatherOfLegends) [inline packages/golden_toolkit/lib/src/multi_screen_golden.dart:98]: @coreysprague I agree we need to come up with a more concise API that pulls this functionality together. We have two interesting behaviors here. One is putting multiple "devices" in one golden, the other is giving each device in the golden unique or different states.

For the first behavior - this is an assumption but, I think if a team or individual is wanting large goldens like this that they probably want all of them like that. If that is true we might be able to make this a configuration. We could also add a way to override this behavior on a case by case basis. I am thinking of it as more of a capture mode.

For the second behavior - assuming we had a capture mode above... I think we could have an optional argument `deviceBuilder` that is mutually exclusive too `devices`. 

@moonytoes29 not asking you to make all these changes but I wanted to include you in the conversation regarding how we can improve the libraries API overall.

File: packages/golden_toolkit/test/device_builder_test.dart [hunk 13]
```diff
@@ -0,0 +1,255 @@
...
+            )
+          ],
+        ),
+      );
+    });
+
+    test('addScenario - overrideDevices - add name', () {
+      // given
+      const deviceName = 'deviceName';
+      final sut = DeviceBuilder();
```
Comment (fatherOfLegends) [inline packages/golden_toolkit/test/device_builder_test.dart:69]: What does `sut` here mean? scenario under test?

File: packages/golden_toolkit/example/lib/src/weather_widgets.dart (hunk 10 above)
Comment (fatherOfLegends) [reply packages/golden_toolkit/example/lib/src/weather_widgets.dart:11]: I have seen the coverage issue with const constructors, I believe we have some tooling that ignores those.

File: packages/golden_toolkit/lib/src/multi_screen_golden.dart (hunk 12 above)
Comment (moonytoes29) [reply packages/golden_toolkit/lib/src/multi_screen_golden.dart:98]: @fatherOfLegends  - this `multiDeviceGolden` is definitely more of a helper for single scenario DeviceBuilder. Doesnt do much extra really. It was really to help me in my current project do a more simple find a replace rather than moving to the DeviceBuilder structure everywhere we already had a multiScreenGolden.

I think if we're looking for a more concise API in the future, we should expand more in the `GoldenBuilder`. I kept this separate as a `DeviceBuilder` to try and not to step on any toes. It could be another factory method under the `GoldenBuilder` namespace.

Another way i thought to tackle this whole PR, was to not create a large widget to encompass all the device sizes, but to just create the option to run any of the current golden methods in a method that just stitches the currently outputted pngs together into one png. But that seemed a lot more complicated.

File: packages/golden_toolkit/README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (moonytoes29) [reply packages/golden_toolkit/README.md]: removed.

File: packages/golden_toolkit/example/lib/src/weather_widgets.dart (hunk 9 above)
Comment (moonytoes29) [reply packages/golden_toolkit/example/lib/src/weather_widgets.dart:9]: removed. will resolve.

Comment (moonytoes29) [reply packages/golden_toolkit/example/lib/src/weather_widgets.dart:9]: removed - will resolve

File: packages/golden_toolkit/lib/src/device_builder.dart (hunk 3 above)
[the remark this answers is not in the document]
Comment (moonytoes29) [reply packages/golden_toolkit/lib/src/device_builder.dart]: Changed to onCreate. Resolve if you guys are happy @coreysprague @fatherOfLegends

File: packages/golden_toolkit/example/test/flutter_demo_page_test.dart (hunk 11 above)
Comment (moonytoes29) [reply packages/golden_toolkit/example/test/flutter_demo_page_test.dart]: So the analysis options succeeds locally, but your pipeline just failed with:
`error • Target of URI doesn't exist: 'package:example/example.dart' •`

So i think i'm going to have to shift back to relative imports to fix this. @fatherOfLegends

**Review by coreysprague — APPROVED** [review]
👍 this looks good to me. I'll wait to merge until @fatherOfLegends agrees

**Review by fatherOfLegends — APPROVED** [review]
👍 thanks for your patience!

**Comment by coreysprague** [comment]
Agreed. Thanks @moonytoes29! I'll update the release notes and publish a new version. Thanks again for your contribution 😄

**Comment by moonytoes29** [comment]
Thanks guys! For all your your feedback. 😁
